File: internal/runtime/lock.go
package runtime

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"
)

type lockRecord struct {
	Owner     string `json:"owner"`
	PID       int    `json:"pid"`
	StartedAt int64  `json:"started_at"`
	Heartbeat int64  `json:"heartbeat"`
}

func acquireWorkspaceLock(workspace, owner string) (func(), error) {
	lockPath := filepath.Join(workspace, ".feng", "lock")
	if err := os.MkdirAll(filepath.Dir(lockPath), 0o755); err != nil {
		return nil, err
	}
	now := time.Now().Unix()
	record := lockRecord{Owner: owner, PID: os.Getpid(), StartedAt: now, Heartbeat: now}
	content, _ := json.MarshalIndent(record, "", "  ")

	for attempts := 0; attempts < 2; attempts++ {
		file, err := os.OpenFile(lockPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o644)
		if err == nil {
			if _, writeErr := file.Write(append(content, '\n')); writeErr != nil {
				_ = file.Close()
				_ = os.Remove(lockPath)
				return nil, writeErr
			}
			if closeErr := file.Close(); closeErr != nil {
				_ = os.Remove(lockPath)
				return nil, closeErr
			}
			markStateLocked(workspace, record)
			stopHeartbeat := startLockHeartbeat(workspace, owner, os.Getpid())
			return func() {
				stopHeartbeat()
				releaseWorkspaceLock(workspace, owner, os.Getpid())
			}, nil
		}
		if !errors.Is(err, os.ErrExist) {
			return nil, err
		}
		if !removeStaleLock(lockPath) {
			return nil, describeLock(lockPath)
		}
	}
	return nil, describeLock(lockPath)
}

func removeStaleLock(lockPath string) bool {
	info, err := os.Stat(lockPath)
	if err != nil {
		return false
	}
	staleAfter := int64(86400)
	if raw := os.Getenv("FENG_LOCK_STALE_SECONDS"); raw != "" {
		if parsed, err := strconv.Atoi(raw); err == nil && parsed > 0 {
			staleAfter = int64(parsed)
		}
	}
	heartbeat := info.ModTime()
	if record, ok := readLockRecord(lockPath); ok && record.Heartbeat > 0 {
		heartbeat = time.Unix(record.Heartbeat, 0)
	}
	if time.Since(heartbeat) < time.Duration(staleAfter)*time.Second {
		return false
	}
	return os.Remove(lockPath) == nil
}

func describeLock(lockPath string) error {
	record, ok := readLockRecord(lockPath)
	if !ok {
		return fmt.Errorf("workspace is locked")
	}
	return fmt.Errorf("workspace is locked by %s pid=%d since=%s heartbeat=%s", record.Owner, record.PID, time.Unix(record.StartedAt, 0).Format(time.RFC3339), time.Unix(record.Heartbeat, 0).Format(time.RFC3339))
}

func releaseWorkspaceLock(workspace, owner string, pid int) {
	lockPath := filepath.Join(workspace, ".feng", "lock")
	data, err := os.ReadFile(lockPath)
	if err == nil {
		var record lockRecord
		if json.Unmarshal(data, &record) == nil && record.Owner == owner && record.PID == pid {
			_ = os.Remove(lockPath)
		}
	}
	_ = updateState(workspace, func(state *State) {
		state.Lock = map[string]string{"owner": "", "heartbeat": ""}
	})
}

func startLockHeartbeat(workspace, owner string, pid int) func() {
	interval := lockHeartbeatInterval()
	stop := make(chan struct{})
	done := make(chan struct{})
	var once sync.Once
	go func() {
		defer close(done)
		ticker := time.NewTicker(interval)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				if !heartbeatWorkspaceLock(workspace, owner, pid) {
					return
				}
			case <-stop:
				return
			}
		}
	}()
	return func() {
		once.Do(func() {
			close(stop)
			<-done
		})
	}
}

func lockHeartbeatInterval() time.Duration {
	raw := strings.TrimSpace(os.Getenv("FENG_LOCK_HEARTBEAT_SECONDS"))
	if raw == "" {
		return 5 * time.Second
	}
	parsed, err := strconv.Atoi(raw)
	if err != nil || parsed < 1 {
		return 5 * time.Second
	}
	if parsed > 3600 {
		parsed = 3600
	}
	return time.Duration(parsed) * time.Second
}

func heartbeatWorkspaceLock(workspace, owner string, pid int) bool {
	lockPath := filepath.Join(workspace, ".feng", "lock")
	record, ok := readLockRecord(lockPath)
	if !ok || record.Owner != owner || record.PID != pid {
		return false
	}
	record.Heartbeat = time.Now().Unix()
	content, _ := json.MarshalIndent(record, "", "  ")
	if err := os.WriteFile(lockPath, append(content, '\n'), 0o644); err != nil {
		return false
	}
	markStateLocked(workspace, record)
	return true
}

func readLockRecord(lockPath string) (lockRecord, bool) {
	data, err := os.ReadFile(lockPath)
	if err != nil {
		return lockRecord{}, false
	}
	var record lockRecord
	if json.Unmarshal(data, &record) != nil {
		return lockRecord{}, false
	}
	return record, true
}

func markStateLocked(workspace string, record lockRecord) {
	_ = updateState(workspace, func(state *State) {
		state.Lock = map[string]string{
			"owner":      record.Owner,
			"pid":        strconv.Itoa(record.PID),
			"heartbeat":  time.Unix(record.Heartbeat, 0).Format(time.RFC3339),
			"started_at": time.Unix(record.StartedAt, 0).Format(time.RFC3339),
		}
	})
}

File: src/feng/lock.py
from __future__ import annotations

import json
import os
import time
from contextlib import contextmanager
from datetime import datetime, timezone
from pathlib import Path
from typing import Iterator

from .state import load_state, save_state
from .utils import ensure_dir


class WorkspaceLocked(RuntimeError):
    pass


def _lock_path(workspace: Path) -> Path:
    return workspace / ".feng" / "lock"


def _rfc3339(ts: int) -> str:
    return datetime.fromtimestamp(ts, tz=timezone.utc).isoformat().replace("+00:00", "Z")


def _stale_after_seconds() -> int:
    raw = os.environ.get("FENG_LOCK_STALE_SECONDS", "").strip()
    if raw:
        try:
            parsed = int(raw)
            if parsed > 0:
                return parsed
        except ValueError:
            pass
    return 86400


def _remove_stale_lock(path: Path) -> bool:
    try:
        age = time.time() - path.stat().st_mtime
    except OSError:
        return False
    if age < _stale_after_seconds():
        return False
    try:
        path.unlink()
        return True
    except OSError:
        return False


def _describe_lock(path: Path) -> str:
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        owner = data.get("owner", "")
        pid = data.get("pid", "")
        started = data.get("started_at", "")
        if owner and pid and started:
            return f"workspace is locked by {owner} pid={pid} since={_rfc3339(int(started))}"
    except Exception:
        pass
    return "workspace is locked"


def _mark_state_locked(workspace: Path, record: dict[str, object]) -> None:
    try:
        state = load_state(workspace)
        state["lock"] = {
            "owner": str(record["owner"]),
            "pid": str(record["pid"]),
            "heartbeat": _rfc3339(int(record["heartbeat"])),
            "started_at": _rfc3339(int(record["started_at"])),
        }
        save_state(workspace, state)
    except Exception:
        pass


def _clear_state_lock(workspace: Path) -> None:
    try:
        state = load_state(workspace)
        state["lock"] = {"owner": "", "heartbeat": ""}
        save_state(workspace, state)
    except Exception:
        pass


@contextmanager
def acquire_workspace_lock(workspace: Path, owner: str) -> Iterator[None]:
    path = _lock_path(workspace)
    ensure_dir(path.parent)
    now = int(time.time())
    pid = os.getpid()
    record: dict[str, object] = {"owner": owner, "pid": pid, "started_at": now, "heartbeat": now}
    for _ in range(2):
        try:
            fd = os.open(path, os.O_CREAT | os.O_EXCL | os.O_WRONLY)
        except FileExistsError:
            if _remove_stale_lock(path):
                continue
            raise WorkspaceLocked(_describe_lock(path))
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            fh.write(json.dumps(record, ensure_ascii=False, indent=2) + "\n")
        _mark_state_locked(workspace, record)
        try:
            yield
        finally:
            _release_workspace_lock(workspace, owner, pid)
        return
    raise WorkspaceLocked(_describe_lock(path))


def _release_workspace_lock(workspace: Path, owner: str, pid: int) -> None:
    path = _lock_path(workspace)
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        if data.get("owner") == owner and int(data.get("pid", -1)) == pid:
            path.unlink()
    except Exception:
        pass
    _clear_state_lock(workspace)
